man/texinfo.5
.\" texinfo(5)
.\"
.\" Copyright 1998-2019  Free Software Foundation, Inc.
.\"
.\" Copying and distribution of this file, with or without modification,
.\" are permitted in any medium without royalty provided the copyright
.\" notice and this notice are preserved.
.\"
.de EX
.nf
.ft CW
.in +5
..
.de EE
.in -5
.ft R
.fi
..
.TH TEXINFO 5 "GNU Texinfo" "FSF"
.SH NAME
texinfo \- software documentation system
.SH DESCRIPTION
Texinfo is a documentation system that uses a single source file to
produce both online information and printed output.  It is primarily
designed for writing software manuals.
.PP
For a full description of the Texinfo language and associated tools,
please see the Texinfo manual (written in Texinfo itself).  Most likely,
running this command from your shell:
.EX
info texinfo
.EE
or this key sequence from inside Emacs:
.EX
M-x info RET m texinfo RET
.EE
will get you there.
.SH AVAILABILITY
http://www.gnu.org/software/texinfo/
.br
or any GNU mirror site.
.SH "REPORTING BUGS"
Please send bug reports to bug-texinfo@gnu.org,
general questions and discussion to help-texinfo@gnu.org.
.SH "SEE ALSO"
\fBinfo\fR(1), \fBinstall-info\fR(1), \fBmakeinfo\fR(1), \fBtexi2dvi\fR(1), \fBtexindex\fR(1).
.br
\fBemacs\fR(1), \fBtex\fR(1).
.br
\fBinfo\fR(5).


NEWS
This NEWS file records noteworthy changes, very tersely.
See the manual for detailed information.

  Copyright 1992-2020 Free Software Foundation, Inc.

  Copying and distribution of this file, with or without modification,
  are permitted in any medium without royalty provided the copyright
  notice and this notice are preserved.

------------------------------------------------------------------------------

6.8 (3 July 2021)
* Language
  . new command @displaymath for formatting of mathematical notation
  . @example takes an argument to specify the language
  . mark these commands as deprecated, not to be used:
    @centerchap, @definfoenclose, @refill, @inforef.
  . new paper size @bsixpaper

* texi2any
  . should be faster as Perl XS parser is enabled by default
  . SHOW_MENU customization variable replaced by FORMAT_MENU.
    FORMAT_MENU set to 'menu' is the same as SHOW_MENU set to 1, and
    FORMAT_MENU set to 'nomenu' is the same as SHOW_MENU set to 0.
  . only check menu structure if CHECK_NORMAL_MENU_STRUCTURE variable is set
  . changes to HTML output:
      . MathJax support for display of math.  new variables HTML_MATH,
        MATHJAX_SCRIPT and MATHJAX_SOURCE.
      . new variables JS_WEBLABELS and JS_WEBLABELS_FILE to support
        JavaScript License Web Labels
      . by default, use sectional tables of contents instead of menus
      . use section names in links by default (configure with
        xrefautomaticsectiontitle customization variable)
      . CONTENTS_OUTPUT_LOCATION sets location of table of contents
      . document sections wrapped in <div> elements
      . new variable USE_NODE_DIRECTIONS to use node or section structure
        for node directions
      . copiable anchor links for definitions with COPIABLE_ANCHORS variable
      . experimental JavaScript browsing interface enabled with INFO_JS_DIR
      . don't add an extra period before file extension given as an argument
        to @image if image file is not found

* info
  . support compressed dir files

* texi2dvi
  . stop on first error in input file

* texinfo.tex
  . put logical page numbers into PDF's ('page labels')
  . put chapter numbers in the PDF outline
  . new Finnish translation

* Distribution
  . autoconf 2.71, automake 1.16.3, gettext 0.21


6.7 (23 September 2019)
* Language:
  . support of index subentries and sub-subentries with @subentry
  . new commands @seeentry and @seealso in index entries
  . no need to wrap Top node in @ifnottex - omitted automatically when
    processed with TeX
  . UTF-8 is the default input encoding

* texi2any
  . for HTML output, mark index nodes in menus and tables of contents
    with the 'rel' attribute of the 'a' tag.
  . TOP_NODE_UP is now only used in HTML if TOP_NODE_UP_URL is set.
    Also TOP_NODE_UP should now be formatted in the output format.
    In HTML TOP_NODE_UP should be suitable for inclusion in HTML 
    element attributes, so for instance should not contain elements.
  . support of noderename.cnf files has been removed
  . INPUT_PERL_ENCODING, INPUT_ENCODING_NAME, NODE_FILE_EXTENSION,
    NODE_FILENAMES, SHORTEXTN and TOP_NODE_FILE removed as customization
    variables.
  . TOP_NODE_FILE_TARGET now contains the extension.
  . error messages translated when the XS parser module is in use

* texi2dvi
  . unconditionally run in --batch mode, i.e. without stopping if there
    is a TeX error
  . keep on going after a TeX error if the index files changed
  . with --tidy (or --build-dir), avoid reading index files from previous
    runs where --tidy was not used

* info
  . for a tree search (with M-/), '}' and '{' work as well as 'M-}' and
    'M-{' to go through the results

* Distribution:
  . Several obsolete portability checks removed
  . gettext 0.20.1, automake 1.16.1


6.6 (16 February 2019)
* Language:
  . new commands @&, @ampchar{}
  . @cropmarks command removed
  . @ctrl is no longer recognised (it was a way to insert literal
    control characters in Info files, but deprecated since the 
    time of Texinfo version 2)
  . \usebracesinindexestrue is no longer recommended for using braces in
    index entries, and has been a no-op for some time

* texi2any
  . extension modules fixed to work with the "thread-safe locales" of
    Perl 5.28 and newer
  . some code changed to stop warnings being given by newer versions of Perl
  . for HTML output, use `id' to define link targets instead of the `name'
    attribute on <a>
  . A native-code implementation of the Texinfo parser has been included
    on an experimental basis, which makes texi2any a lot faster.  Set the
    `TEXINFO_XS_PARSER' environment variable to 1 to use.
  . changes to HTML output:
     . omit colon after node name in menus by default (use
       `MENU_ENTRY_COLON' to add it back)
     . no special CSS for commands like @smallexample
     . new customization variable `SECTION_NAME_IN_TITLE' to use the
       section name as the document <title>
     . use section names instead of node names in generated menus
  . pass on flags set with -D to TeX
  . useless static libraries are not installed
  . the newline after an @insertcopying is not output
  . warning given for @multitable prototypes not in braces
  . @indent and @noindent are not allowed inside the arguments to
    commands where they are not meaningful
  . @quote-arg and @allow-recursion are not recognised (these two used
    to be recognised by makeinfo in macro definitions but were never 
    implemented in texinfo.tex)
  . `FIX_TEXINFO' removed as a customization variable
  . do not recognise or warn about obsolete customization variables

* info
  . debugging output with -x is not diverted to a separate infodebug file

* Development:
  . switch from Subversion to git
     - https://savannah.gnu.org/git/?group=texinfo
  . automake 1.16


6.5 (12 September 2017)
* info:
  . some bugs fixed:
    . a bug where a segfault could happen in the regex search, for

tp/Texinfo/Convert/Texinfo.pm
# Texinfo.pm: output a Texinfo tree as Texinfo.
#
# Copyright 2010-2022 Free Software Foundation, Inc.
# 
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License,
# or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
# 
# Original author: Patrice Dumas <pertusus@free.fr>
# Parts (also from Patrice Dumas) come from texi2html.pl or texi2html.init.

package Texinfo::Convert::Texinfo;

use 5.00405;
use strict;

# To check if there is no erroneous autovivification
#no autovivification qw(fetch delete exists store strict);

use Carp qw(cluck confess);

# commands definitions
use Texinfo::Commands;

require Exporter;
use vars qw($VERSION @ISA @EXPORT_OK %EXPORT_TAGS);
@ISA = qw(Exporter);

%EXPORT_TAGS = ( 'all' => [ qw(
  convert_to_texinfo
  node_extra_to_texi
) ] );

@EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

$VERSION = '7.0';


my %brace_commands           = %Texinfo::Commands::brace_commands;
my %block_commands           = %Texinfo::Commands::block_commands;
my %def_commands             = %Texinfo::Commands::def_commands;

# used in root_heading_command_to_texinfo
my %sectioning_heading_commands = %Texinfo::Commands::sectioning_heading_commands;

my @ignored_types = ('spaces_inserted', 'bracketed_inserted',
'command_as_argument_inserted');
my %ignored_types;
for my $a (@ignored_types) {
  $ignored_types{$a} = 1;
}

# used to put a node name in error messages.
sub node_extra_to_texi($)
{
  my $node = shift;
  my $result = '';
  if ($node->{'manual_content'}) {
    $result = '('.convert_to_texinfo({'contents'
                              => $node->{'manual_content'}}) .')';
  }
  if ($node->{'node_content'}) {
    $result .= convert_to_texinfo({'contents' => $node->{'node_content'}});
  }
  return $result;
}

# for debugging.
sub root_heading_command_to_texinfo($)
{
  my $element = shift;
  my $tree;
  if ($element->{'cmdname'}) {
    if ($element->{'cmdname'} eq 'node') {
      $tree = $element->{'extra'}->{'node_content'};
    } elsif ($sectioning_heading_commands{$element->{'cmdname'}}
             and $element->{'args'}->[0]->{'contents'}) {
      $tree = $element->{'args'}->[0]->{'contents'};
    }
  } else {
    return "Not a root command";
  }
  return '@'.$element->{'cmdname'}.' '.convert_to_texinfo({'contents' => $tree})
          if ($tree);
  return '@'.$element->{'cmdname'};
}

# Following subroutines deal with transforming a texinfo tree into texinfo
# text.  Should give the text that was used parsed, except for a few cases.

# the second arguments, if defined triggers replaced
# tree item to be shown, in the default case they are
# not shown.
# expand a tree to the corresponding texinfo.
sub convert_to_texinfo($;$);
sub convert_to_texinfo($;$)
{
  my $element = shift;
  my $expand_replaced = shift;

  confess "convert_to_texinfo: element undef" if (!defined($element));
  confess "convert_to_texinfo: bad element type (".ref($element).") $element"
     if (ref($element) ne 'HASH');
  my $result = '';

  return '' if ($element->{'type'}
                and ($ignored_types{$element->{'type'}}
                     or ($element->{'type'} eq 'replaced'
                         and not $expand_replaced)));
  if (defined($element->{'text'})) {
    $result .= $element->{'text'};
  } else {
    if ($element->{'cmdname'}
       or ($element->{'type'} and ($element->{'type'} eq 'def_line'
                                or $element->{'type'} eq 'menu_entry'
                                or $element->{'type'} eq 'menu_comment'))) {
      $result .= _expand_cmd_args_to_texi($element, $expand_replaced);
    }
    if ($element->{'type'}
        and ($element->{'type'} eq 'bracketed'
             or $element->{'type'} eq 'bracketed_def_content')) {
      $result .= '{';
      if ($element->{'extra'}
          and $element->{'extra'}->{'spaces_before_argument'}) {
         $result .= $element->{'extra'}->{'spaces_before_argument'};
      }
    }
    if (defined($element->{'contents'})) {
      foreach my $child (@{$element->{'contents'}}) {
        $result .= convert_to_texinfo($child, $expand_replaced);
      }
    }
    if ($element->{'extra'} and $element->{'extra'}->{'spaces_after_argument'}) {
      $result .= $element->{'extra'}->{'spaces_after_argument'};
    }
    if ($element->{'extra'} and $element->{'extra'}->{'comment_at_end'}) {
      $result .= convert_to_texinfo($element->{'extra'}->{'comment_at_end'},
                         $expand_replaced);
    }
    $result .= '}' if ($element->{'type'}
                       and ($element->{'type'} eq 'bracketed'
                            or $element->{'type'} eq 'bracketed_def_content'));
  }
  return $result;
}

# expand a command argument as texinfo.
sub _expand_cmd_args_to_texi($;$) {
  my $cmd = shift;
  my $expand_replaced = shift;

  my $cmdname = $cmd->{'cmdname'};
  $cmdname = '' if (!$cmd->{'cmdname'});
  my $result = '';
  $result = '@'.$cmdname if ($cmdname);

  # this is done here otherwise for some constructs, there are
  # no 'args', and so the space is never readded.
  if ($cmd->{'extra'} and exists ($cmd->{'extra'}->{'spaces'})) {
    $result .= $cmd->{'extra'}->{'spaces'};
  }
  # block line commands with arguments not separated by commas
  if ($block_commands{$cmdname}
         and ($def_commands{$cmdname}
              or $block_commands{$cmdname} eq 'multitable')
         and $cmd->{'args'}) {
     $result .= $cmd->{'extra'}->{'spaces_before_argument'}
       if $cmd->{'extra'} and $cmd->{'extra'}->{'spaces_before_argument'};
     foreach my $arg (@{$cmd->{'args'}}) {
        $result .= convert_to_texinfo($arg, $expand_replaced);
    }
  # arg_line set for line_commands with type special
  } elsif (($cmd->{'extra'} or $cmdname eq 'macro' or $cmdname eq 'rmacro')
           and defined($cmd->{'extra'}->{'arg_line'})) {
    $result .= $cmd->{'extra'}->{'spaces_before_argument'}
      if $cmd->{'extra'} and $cmd->{'extra'}->{'spaces_before_argument'};
    $result .= $cmd->{'extra'}->{'arg_line'};
  } elsif (($block_commands{$cmdname} or $cmdname eq 'node')
            and defined($cmd->{'args'})) {
    $result .= $cmd->{'extra'}->{'spaces_before_argument'}
      if $cmd->{'extra'} and $cmd->{'extra'}->{'spaces_before_argument'};
    foreach my $arg (@{$cmd->{'args'}}) {
      next if $arg->{'type'} and $ignored_types{$arg->{'type'}};
      if ($arg->{'extra'} and $arg->{'extra'}->{'spaces_before_argument'}) {
        $result .= $arg->{'extra'}->{'spaces_before_argument'};
      }
      $result .= convert_to_texinfo($arg);
      $result .= ',';
    }
    $result =~ s/,$//;
  } elsif (defined($cmd->{'args'})) {
    my $braces;
    $braces = 1 if ($cmd->{'args'}->[0]->{'type'}
                    and ($cmd->{'args'}->[0]->{'type'} eq 'brace_command_arg'
                         or $cmd->{'args'}->[0]->{'type'} eq 'brace_command_context'));
    $result .= '{' if ($braces);
    if ($cmdname eq 'verb') {
      $result .= $cmd->{'extra'}->{'delimiter'};
    }
    if ($cmd->{'extra'}
        and $cmd->{'extra'}->{'spaces_before_argument'}) {
      $result .= $cmd->{'extra'}->{'spaces_before_argument'};
    }
    my $arg_nr = 0;
    foreach my $arg (@{$cmd->{'args'}}) {
      if (exists($brace_commands{$cmdname}) or ($cmd->{'type'}
                    and $cmd->{'type'} eq 'definfoenclose_command')) {
        $result .= ',' if ($arg_nr);
        $arg_nr++;
      }
      if ($arg->{'extra'} and $arg->{'extra'}->{'spaces_before_argument'}) {
        $result .= $arg->{'extra'}->{'spaces_before_argument'};
      }
      $result .= convert_to_texinfo($arg);
    }
    if ($cmdname eq 'verb') {
      $result .= $cmd->{'extra'}->{'delimiter'};
    }
    $result .= '}' if ($braces);
  } else {
    $result .= $cmd->{'extra'}->{'spaces_before_argument'}
      if $cmd->{'extra'} and $cmd->{'extra'}->{'spaces_before_argument'};
  }
  $result .= '{'.$cmd->{'extra'}->{'flag'}.'}' if ($cmdname eq 'value');
  return $result;
}

1;
__END__

=head1 NAME

Texinfo::Convert::Texinfo - Convert a Texinfo tree to Texinfo code

=head1 SYNOPSIS

  use Texinfo::Convert::Texinfo qw(convert_to_texinfo);
  
  my $texinfo_text = convert_to_texinfo($tree);

=head1 NOTES

The Texinfo Perl module main purpose is to be used in C<texi2any> to convert
Texinfo to other formats.  There is no promise of API stability.

=head1 DESCRIPTION

C<Texinfo::Convert::Texinfo> converts a Texinfo tree (described in
L<Texinfo::Parser>) to Texinfo code.  If the Texinfo tree results from
parsing some Texinfo document, The converted Texinfo code should be
exactly the same as the initial document, except that user defined @-macros
and C<@value> are expanded, and some invalid code is discarded.

=head1 METHODS

=over

=item $texinfo_text = convert_to_texinfo($tree)
X<C<convert_to_texinfo>>

Converts the Texinfo tree I<$tree> to Texinfo code.

=back

=head1 AUTHOR

Patrice Dumas, E<lt>pertusus@free.frE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2010- Free Software Foundation, Inc.  See the source file for
all copyright years.

This library is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 3 of the License, or (at
your option) any later version.

=cut
